@W-18172401 Integrate getUrlMapping on server-side navigation

hajinsuha1 opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->

# Description

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- add `extension-seo-url-mapping` extension
  - overrides the `extendRoutes` method to make a Shopper SEO `getUrlMapping` API call and add it to the routes if a response is returned
- update `pwa-kit-react-sdk` to serialize and deserialize routes to allow dynamically modifying the routes list

# How to Test-Drive This PR

- Revert this commit to test the changes
    ```
    git revert 98a5011bfeafe4aa66df8c34983afa1745690cf3
    ```
- Build the app
    ```
    npm install
    ```
- Start the local server
    ```
    cd `packages/template-typescript-minimal`
    npm start
    ```
- Navigate to http://localhost:3000/sample-page
  - verify the Sample page is rendered with title "A custom title from the extension"
- Navigate to http://localhost:3000/seo-category
  - verify a PLP page with the `New Arrivals` category is rendered
- Navigate to http://localhost:3000/seo-product
  - verify a PDP page with the `Platinum Blue Stripes Easy Care Fitted Shirt` product is rendered
- Navigate to http://localhost:3000/seo-redirect
  - verify that you are redirected to `/category/top-sellers`

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

hajinsuha1 committed:
initial implementation of extension seo url mapping extension

hajinsuha1 committed:
make beforeRouteMatch async

hajinsuha1 committed:
generate component map from react-sdk util function

hajinsuha1 commented on packages/extension-seo-url-mapping/src/setup-app.ts:
```diff
@@ -0,0 +1,110 @@
+        let {resourceTypeToComponentMap} = config
+        console.log('config', config, 'resourceTypeToComponentMap', resourceTypeToComponentMap)
+        if (resourceTypeToComponentMap === undefined) {
+            resourceTypeToComponentMap = generateResourceTypeMap(allRoutes)
```
It seems like callback functions are not supported values for the `config/default.js`. Config fields with a callback function as the value get passed as undefined. 

As a temporary workaround, I've copied the config code to a `generateResourceTypeMap ` util function in the sdk.

hajinsuha1 committed:
use a component map that maps resource type to component display name

hajinsuha1 committed:
move logic from beforeRouteMatch to extendRoutes

hajinsuha1 committed:
make extendRoutes async

hajinsuha1 committed:
pass request to extendRoutes

hajinsuha1 committed:
add route serialization

hajinsuha1 committed:
make category urlMappings work

hajinsuha1 committed:
make urlMapping resourceType product work

hajinsuha1 committed:
cleanup

hajinsuha1 committed:
move serializing to pwa-kit-react-sdk

hajinsuha1 commented on packages/extension-seo-url-mapping/src/setup-app.tsx:
```diff
@@ -0,0 +1,110 @@
+    async extendRoutes(routes: RouteProps[], req?: Request): Promise<RouteProps[]> {
+        const path = req?.path
+        const {resourceTypeToComponentMap} = this.getConfig()
+        // TODO: How to call urlMapping API when React Hook "useUrlMapping"
+        // cannot be called in a class component. React Hooks must be called
+        // in a React function component or a custom React Hook function.
+        const mapping = await getUrlMapping(path)
```
@bendvc I think you briefly mentioned this but the `useUrlMapping` hook from `commerce-sdk-react` can't be called here because it needs to be used within a React hook or component. 

For now instead of making an actual API call I made a mock function `getUrlMapping` that returns the response for the following types of urlMappings: `product`, `category`, `redirect`, `content_asset`

bendvc replied:
Please reach out the extensibility channel so we can start brain storming some idea's to solve this problem. I think this is potentially a blocker to getting this implemented.

hajinsuha1 commented on packages/pwa-kit-react-sdk/src/ssr/browser/main.jsx:
```diff
@@ -132,10 +132,40 @@ export const start = async () => {
         locals
     })
 
+    // Await the async getRoutes function
+    let routes = await getRoutes(locals)
+
+    let serializedRoutes = window.__CONFIG__.app.routes
+    // Deserialize routes
+    serializedRoutes = serializedRoutes.map(
+        ({path, componentName, componentProps}) => {
+            let component = routes.find((route) =>
+                route.component?.displayName?.includes(componentName)
+            )?.component
+            if (!component) {
+                // TODO: Error handling if given component couldn't be found
+                console.error('Component', componentName, 'could not be deserialized for path', path)
+                return
+            }
+
+            if (componentProps) {
+                const Component = component
+                component = () => <Component {...componentProps} />
+            }
+            return {
+                path,
+                exact: true,
+                component
+            }
+        }
+    )
+    serializedRoutes = serializedRoutes.filter((route) => !!route)
+    routes = serializedRoutes
```
@bendvc Moved the deserialization to here so it only runs on client-side

bendvc replied:
Interesting.. would the async logic for making your seo url mapping called be called now on the client even though you have the routes serialized. 

E.g. on line 135 you get the routes which will call the extendRoutes for all extensions and await on it.. then you will deserialize the routes starting at line 138. Effectively coming up with the same route list?

hajinsuha1 replied:
Good point it does seem unnecessary to deserialize routes that are already in the routes array. 

I'll look to only serialize/deserialize the route of the current path

hajinsuha1 commented on packages/pwa-kit-react-sdk/src/utils/routes.js:
```diff
@@ -0,0 +1,42 @@
+            to: urlMapping.destinationUrl
+        }
+    } else {
+        // TODO: support "content_asset" resource types
+        Component = component
+        props = {
+            [`${urlMapping.resourceType}Id`]: urlMapping.resourceId
+        }
```
@bendvc still need to add support for mapping `content_asset` resource types to a component

hajinsuha1 commented on packages/pwa-kit-react-sdk/src/utils/routes.js:
```diff
@@ -0,0 +1,42 @@
+
+    return {
+        path: path,
+        // DEVELOPER NOTE: Here we would want to use a Loadable component as to not bloat the home page chunk size.
+        component: Component,
```
@bendvc this note was here in the spike but do we still want to wrap this with a `Loadable` component?

bendvc commented on packages/extension-seo-url-mapping/README.md:
Are you planning on updating the README in this ticket or splitting it out into it's own?

hajinsuha1 replied:
I'm thinking it'll be better to be on it's own so that we can write the README after we have all the changes for this extension implemented

bendvc commented on packages/extension-seo-url-mapping/config/README.md:
This file isn't needed for our extensions and it's really only used for customers to help them develop their extension. So you can delete it.

bendvc commented on packages/extension-seo-url-mapping/package.json:
```diff
@@ -0,0 +1,45 @@
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0"
+  },
+  "peerDependenciesMeta": {
+    "@salesforce/pwa-kit-extension-sdk": {
+      "optional": true
+    },
+    "@salesforce/pwa-kit-react-sdk": {
+      "optional": true
+    }
+  },
```
These were removed from the sample extension code (or they should have been) and shouldn't be in your extension. you can remove them.

bendvc commented on packages/extension-seo-url-mapping/src/setup-app.tsx:
```diff
@@ -0,0 +1,110 @@
+// Others
+import extensionMeta from '../extension-meta.json'
+
+interface StoreSlice {
+    count: number
+    increment: () => void
+    decrement: () => void
+}
+
+// This is safe to delete if your extension does not use state. If you aren't using this, ensure you remove the
+// `withApplicationExtensionStore` usage below as well.
+const sliceInitializer: SliceInitializer<StoreSlice> = (set) => ({
+    count: 0,
+    increment: () => set((state) => ({count: state.count + 1})),
+    decrement: () => set((state) => ({count: state.count - 1}))
+})
```
If you aren't going to be using a global store, you can remove this.

bendvc commented on packages/extension-seo-url-mapping/src/setup-app.tsx:
```diff
@@ -0,0 +1,110 @@
+        App: React.ComponentType<T>
+    ): React.ComponentType<T> {
+        const HOCs = [
+            // Optionally include state for this extension using `withApplicationExtensionStore`
+            (component: React.ComponentType<any>) =>
+                withApplicationExtensionStore(component, {
+                    id: extensionMeta.id,
+                    initializer: sliceInitializer
+                })
```
This too.

bendvc commented on packages/extension-seo-url-mapping/src/types/config.ts:
```diff
@@ -0,0 +1,38 @@
+
+// Defines the map of resource types to component display name.
+// https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-seo?meta=getUrlMapping
+type ResourceTypeToComponentMap = {
+    category: string
+    product: string
+    content_asset: string
+}
```
We might want to leverage the enum that the API uses or the response type and gave those values.. we know that the key might be an entry of the enum = CATEGORY | PRODUCT | CONTENT_ASSET which would hep clean up this type.

bendvc replied:
We also still have the problem where we are mapping to the components display name. So we'll want to harden that solution a little. Maybe the pages in the storefront and store locator need to be prefixed with the extensions name/id to help prevent collisions or duplicate component/page names which might cause this solution to fall flat.

bendvc commented on packages/extension-seo-url-mapping/src/setup-app.tsx:
```diff
@@ -0,0 +1,110 @@
+                mapping.resourceType as keyof typeof resourceTypeToComponentMap
+            ]
+        // TODO error handling if component not found
+        const component = routes.find((route) =>
```
One of the caveats of using `extendRoutes`, and I know that I suggested that is the api method we use for this, is that "routes" is only a list of routes from the extensions that have been applied to the application before it. 

Customers will have to know this via documentation (readme) that they have to add the seo mapping extension after all the extension, which might be weird. So lets put that in a decisions or findings section in your PR

bendvc commented on packages/extension-seo-url-mapping/src/types/index.ts:
```diff
@@ -0,0 +1,8 @@
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+export * from './config'
```
You'll want to export your serialize route type here as well.

bendvc commented on packages/extension-seo-url-mapping/src/types/routes.ts:
Let go with the patter that each type has it's own file and the types file name and type name are synced. So we should call this serializedRoute.ts

hajinsuha1 replied:
after I moved the serialization/deserialization to `pwa-kit-react-sdk` this type is no longer in use so i've deleted it!

bendvc commented on packages/pwa-kit-react-sdk/src/ssr/browser/main.jsx:
I think it will be wise to break out the reach sdk changes for serialization of the routes into it's own PR and then we can easily share that with more pwa-kit team members.

hajinsuha1 replied:
Moved to https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2300

bendvc commented on packages/pwa-kit-react-sdk/src/ssr/universal/components/route-component/index.js:
```diff
@@ -402,17 +402,17 @@ export const routeComponent = (Wrapped, isPage, locals) => {
-    applicationExtensions.forEach((applicationExtension) => {
-        _routes = applicationExtension.extendRoutes(_routes)
-    })
+    for (const applicationExtension of applicationExtensions) {
+        _routes = await applicationExtension.extendRoutes(_routes, req)
+    }
```
Will this code run the promises in parallel or one after another?

hajinsuha1 replied:
This code will run one after another

bendvc commented on packages/template-typescript-minimal/package.json:
```diff
@@ -39,7 +39,9 @@
-    "zustand": "5.0.3"
+    "zustand": "5.0.3",
+    "@salesforce/extension-chakra-storefront": "0.1.0-extensibility-preview.4",
+    "@salesforce/extension-seo-url-mapping": "0.0.1"
```
You might find that after you merge the latest work from the feature branch into yours that this configured typescript minimal app doesn't work anymore. Thats because you have to manually install the extensions peer dependancies by hand (you only have to do this if you are working in a mono-repo template project). So you'll have to copy the peer dependencies from each extension and add them here, ensuring you remove duplicates and update some of the mono-repo package versions.

hajinsuha1 committed:
move transformUrlMappingToRoute from pwa-kit-react-sdk to extension-seo-url-mapping

hajinsuha1 committed:
Merge branch 'feature/extensibility-v2' into implement-dynamic-update-routes

hajinsuha1 committed:
cleanup

hajinsuha1 committed:
update package.json and package-lock.json in template-typescript-minimal

hajinsuha1 committed:
attempt to fix component not loading with props on server side

hajinsuha1 committed:
implement route serialization/deserialization

hajinsuha1 committed:
Merge branch 'route-serialization-to-support-shopper-seo-integration' into implement-dynamic-update-routes

hajinsuha1 committed:
initial pass at implementing route serialization via a componentMap

hajinsuha1 committed:
update error handling message format

hajinsuha1 committed:
Merge branch 'route-serialization-to-support-shopper-seo-integration' into implement-dynamic-update-routes

hajinsuha1 committed:
cleanup and fix getRoutes to not duplicate routes

vmarta committed:
Convert extendRoutes to an async method

hajinsuha1 committed:
create a helper function generateComponentMapFromModules

vmarta committed:
Add todos

vmarta committed:
Temporarily modify the typescript-minimal project

hajinsuha1 committed:
initial impl of adding base template components to component map

vmarta committed:
Minor refactoring

vmarta committed:
Call every extension's route function in parallel

For a simpler implementation, replaced extendRoutes with getRoutes. Probably safer too, since we think there's no need to have access to the existing/previous routes.

hajinsuha1 committed:
create a prefixDisplayName utils

hajinsuha1 committed:
update error handling

hajinsuha1 committed:
fix deserialization of components from pwa-kit-react-sdk

hajinsuha1 committed:
Merge branch 'vm/async-extendRoutes' into route-serialization-to-support-shopper-seo-integration

hajinsuha1 committed:
merge

hajinsuha1 committed:
initial implementation of serialize/deserialize

hajinsuha1 committed:
cleanup

hajinsuha1 committed:
make serialize() and getComponentMap() synchronous

vmarta committed:
Pass in the request path

hajinsuha1 committed:
cache getRoutes using CacheResult decorator

hajinsuha1 committed:
move deserialization logic to constructor of ApplicationExtension

hajinsuha1 committed:
call this.getComponentMap within deserialize and fix typing

hajinsuha1 committed:
cleanup serialize logic

vmarta committed:
Explore passing in locals and how to manage access token

vmarta committed:
Remove extendRoutes from the API

vmarta committed:
Allow configuring of the auth property name in locals

hajinsuha1 committed:
comments + create DeserializedExtension type

hajinsuha1 committed:
make getComponentMap abstract

hajinsuha1 committed:
add ComponentMap type to getComponentMap method in ChakraStorefront and Sample classes

hajinsuha1 committed:
add displayName to StoreLocatorPage

vmarta committed:
Rename the property in locals

hajinsuha1 committed:
add unit tests for cacheResult

vmarta committed:
Create a new extension for a better example

vmarta committed:
Add some console logs

hajinsuha1 committed:
refactor Switch component to use RoutesProvider and create RoutesConsumer for better route management

hajinsuha1 committed:
update CacheResult decorator to support async methods and modify tests accordingly

hajinsuha1 committed:
enhance CacheResult decorator to support both async and sync methods and update tests

hajinsuha1 committed:
refactor getRoutes method to support both sync and async returns and update related logic

hajinsuha1 committed:
refactor ApplicationExtension to enforce getComponentMap() for async getRoutes

hajinsuha1 committed:
cleanup ApplicationExtension and apply CacheResult to getRoutes method in constructor

hajinsuha1 committed:
move CacheResult decorator logic to applyCacheForMethod in ApplicationExtension

hajinsuha1 committed:
don't make getComponentMap abstract but make it optional

hajinsuha1 committed:
remove unused ComponentMap import from test files

vmarta committed:
Experiment with having partial route

Partial route in the getRoutes, and then completing it during beforeRouteMatch.

vmarta committed:
Experiment with correct way to get app origin

vmarta committed:
Add todo

vmarta committed:
Some cleanup

hajinsuha1 committed:
lint

hajinsuha1 committed:
cleanup

vmarta committed:
Fix the partial route with correct path

vmarta committed:
Remove no longer necessary code

vmarta committed:
Combine params into a single object

vmarta committed:
Remove some console logs

hajinsuha1 committed:
add async route handling and serialization tests for ApplicationExtension

vmarta committed:
Add comment to clarify

hajinsuha1 committed:
handle all condiiton in deserialize and move applyCacheForMethod to a utils

hajinsuha1 committed:
refactor: rename getRoutes to getRoutesAsync for clarity and update related references

hajinsuha1 committed:
refactor: remove isGetRoutesAsync method as it's no longer needed

vmarta committed:
Implement getRoutes and getRoutesAsync

Co-Authored-By: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>
Co-Authored-By: Ben Chypak <bchypak@salesforce.com>

vmarta committed:
Add comment

vmarta committed:
Add comment

hajinsuha1 committed:
Merge branch 'vm/async-extendRoutes' into route-serialization-to-support-shopper-seo-integration

Signed-off-by: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>

hajinsuha1 committed:
update getRoutesAsync to also return SerializedRoute and serialize to accept SerializedRoute

hajinsuha1 committed:
cleanup decorator and chakra-storefront

hajinsuha1 committed:
update URLMapping.getRoutesAsync and BeforeRouteMatchParams to support SerializedRoute

hajinsuha1 committed:
address PR feedback

hajinsuha1 committed:
update error message

hajinsuha1 committed:
fix cacheMethodResult

vmarta committed:
Add type to allow for `componentName` property

hajinsuha1 committed:
fix ApplicationExtension.test.tsx

hajinsuha1 committed:
Merge branch 'vm/async-extendRoutes' into route-serialization-to-support-shopper-seo-integration

Signed-off-by: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>

vmarta committed:
Remove params for `getRoutes`

Only `getRoutesAsync` who needs the params.

vmarta committed:
Remove params for getRoutes

hajinsuha1 committed:
refactor: update route handling to use RouteProps exclusively and improve error messages

hajinsuha1 committed:
Merge branch 'vm/async-extendRoutes' into route-serialization-to-support-shopper-seo-integration

Signed-off-by: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>

hajinsuha1 committed:
refactor: rename GetRoutesParams to GetRoutesAsyncParams for to fix error

hajinsuha1 committed:
update unit tests

vmarta committed:
Revert "Remove params for `getRoutes`"

This reverts commit 38e352ffeee0abe2fa42f461177e091d9fabcea2.

vmarta committed:
Revert "Remove params for getRoutes"

This reverts commit 4119cda0fc8474bdfde64ce22a7a9af2e17667e6.

hajinsuha1 committed:
Merge branch 'vm/async-extendRoutes' into route-serialization-to-support-shopper-seo-integration

Signed-off-by: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>

hajinsuha1 committed:
reorganize imports for RoutesContext and unit tests

hajinsuha1 committed:
Merge branch 'route-serialization-to-support-shopper-seo-integration' of https://github.com/SalesforceCommerceCloud/pwa-kit into route-serialization-to-support-shopper-seo-integration

hajinsuha1 committed:
Add isServerSide utility function and corresponding tests

hajinsuha1 committed:
lint

hajinsuha1 committed:
[REVERT] temporary logging for testing

hajinsuha1 committed:
Remove unused properties from global Window interface in type definitions

hajinsuha1 committed:
update changelog

vmarta committed:
Refactor so that extension is the one in control of serialization

In other words, future proofing it.

hajinsuha1 committed:
Update packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts

Co-authored-by: Vincent Marta <vmarta@salesforce.com>
Signed-off-by: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>

hajinsuha1 committed:
update unit tests and add error handling

hajinsuha1 committed:
Merge branch 'route-serialization-to-support-shopper-seo-integration' into vm/tweak-serialized-extensions

hajinsuha1 committed:
lint

hajinsuha1 committed:
lint

hajinsuha1 committed:
lint extension-url-mapping

hajinsuha1 committed:
Merge branch 'route-serialization-to-support-shopper-seo-integration' into implement-dynamic-update-routes

vmarta committed:
Update comments

vmarta committed:
Update tests

hajinsuha1 committed:
rename extension-url-mapping to extension-commerce-bm-seo

vmarta committed:
Update tests

vmarta committed:
Revert changes to typescript-minimal

vmarta committed:
Merge branch 'feature/extensibility-v2' into vm/async-extendRoutes

vmarta committed:
Update package-lock.json

vmarta committed:
Update changelog files

vmarta committed:
Update readme files

hajinsuha1 committed:
Merge branch 'vm/async-extendRoutes' into route-serialization-to-support-shopper-seo-integration

Signed-off-by: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>

hajinsuha1 committed:
Refactor ApplicationExtension tests to use correct component types and update changelog for SSR enhancements

hajinsuha1 committed:
Remove extension-url-mapping package and associated files

hajinsuha1 committed:
fix unit tests

hajinsuha1 committed:
Revert "Remove extension-url-mapping package and associated files"

This reverts commit e08d7ce66069ec41cc9615266600636411020cba.

hajinsuha1 committed:
lint

hajinsuha1 committed:
Revert "Revert changes to typescript-minimal"

This reverts commit 7ba4d99c69e67e18df4eb0d148cde36fe405f968.

hajinsuha1 committed:
remove extension-url-mapping, logging, changes to typescript-minimal

This reverts commit 0cd67cc53dfebc3f68472d29ee6b8e6d1931430b.

hajinsuha1 committed:
update maxSize for build/vendor.js to 353 kB in template-retail-react-app

hajinsuha1 committed:
update CHANGELOG.md for template-retail-react-app

hajinsuha1 committed:
add ResourceTypeToComponentMap and move functions to utils

vmarta committed:
Add jsdoc to some types

hajinsuha1 committed:
Update packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts

Co-authored-by: Vincent Marta <vmarta@salesforce.com>
Signed-off-by: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>

hajinsuha1 committed:
address PR feedback (remove unnecessary code)

hajinsuha1 committed:
Revert "remove extension-url-mapping, logging, changes to typescript-minimal"

This reverts commit 54ec28af6d7e08da8b4dcc905c407b14d7166ba9.

hajinsuha1 committed:
update ApplicationExtension unit test

hajinsuha1 committed:
Revert "Revert "remove extension-url-mapping, logging, changes to typescript-minimal""

This reverts commit 876b4c24009cbb0a8152255f07733c178e1dcf89.

hajinsuha1 committed:
Merge branch 'vm/async-extendRoutes' into route-serialization-to-support-shopper-seo-integration

hajinsuha1 committed:
Revert "remove extension-url-mapping, logging, changes to typescript-minimal"

This reverts commit 54ec28af6d7e08da8b4dcc905c407b14d7166ba9.

hajinsuha1 committed:
Merge branch 'route-serialization-to-support-shopper-seo-integration' into implement-dynamic-update-routes

hajinsuha1 committed:
refactor such that extension-commerce-bm-seo handles deserialization at beforeRouteMatch

hajinsuha1 committed:
refactor serialize and deserialize to be serializeRoutes and deserializeRoutes

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/setup-app.ts:
```diff
@@ -0,0 +1,177 @@
+        return result
+    }
+
+    getComponentMap(): ComponentMap {
+        // During deserialize each serialized route in this extension MUST return a component.
+        // This implementation returns an object where the keys are component names in resourceTypeToComponentMap
+        // and the values are placeholder components with the displayName
+        const {resourceTypeToComponentMap} = this.getConfig()
+        return Object.values(resourceTypeToComponentMap).reduce((acc: ComponentMap, name) => {
+            acc[`${this.getName()}.${name}`] = Object.assign(() => null, {displayName: `${this.getName()}.${name}`})
+            return acc
+        }, {} as Record<string, string>)
+    }
```
This implementation returns an object where the keys are component names in `resourceTypeToComponentMap` and the values are placeholder components with the `displayName` in the format `CommerceBmSeo.ProductList`.

This is needed for deserializing the routes component from the server to the client

hajinsuha1 committed:
cleanup beforeRouteMatch

hajinsuha1 commented on packages/extension-starter/src/setup-app.ts:
```diff
@@ -79,14 +79,18 @@ class Sample extends ApplicationExtension<Config> {
-    getRoutes(params: GetRoutesParams): RouteProps[] {
-        return [
+    getRoutesAsync(params: GetRoutesParams): Promise<RouteProps[]> {
+        return Promise.resolve([
             {
                 exact: true,
                 path: this.getConfig().path,
                 component: SamplePage
             }
-        ]
+        ])
+    }
+
+    getComponentMap() {
+        return {SamplePage}
```
to ensure that the solution works for extensions that implement `getRoutesAsync` I have temporarily updated `extension-starter`

hajinsuha1 replied:
`getComponentMap` is required because it is used to deserialize the serialized routes from the server in the client.

Extensions need to have this implemented so that we can map a `componentName` to `component` during deserialization

vmarta replied:
Thanks for making this temporary change to `extension-starter`. I found myself so focused on the SEO extension, that I forgot the possibly-more-common case, which the extension-starter is in.

hajinsuha1 commented on packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts:
```diff
@@ -97,33 +96,33 @@ export class ApplicationExtension<
      * @throws Error if the routes have not been loaded.
      */
-    public serialize(): SerializedExtension | null {
+    public serializeRoutes(): SerializedRouteProps[] | null {
```
since a serialized extension only contains routes i've renamed this method to just `serializeRoutes`.
In the future if we serialize more data we can make a `serialize` method that calls this method.

More details: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2342#discussion_r2021462448

@vmarta I chose to keep it in ApplicationExtension because deserialize needs to be called in the ApplicationExtension's constructor so seemed like it would be better to keep both `serializeRoutes` and `deserializeRoutes` in this class

vmarta replied:
Yeah, I'm ok with having both those methods in the same file (ApplicationExtension).

Aside: because of this [comment thread](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2342/files#r2021294123), I was thinking we actually do need the component map.. and similarly for serialize and deserialize. Anyways, I'm saying that I'm now convinced we need all of them, and I like the direction this PR takes.

hajinsuha1 commented on packages/pwa-kit-extension-sdk/src/types.ts:
```diff
@@ -69,12 +62,11 @@ export type ComponentMap = {
  */
-export type DeserializedExtension = {
-    routes: RouteProps[]
+export type SerializedRouteProps = RouteProps & {
+    componentName?: string
 }
```
introduced a new type `SerializedRouteProps` that is returned by the `serialize` method

hajinsuha1 committed:
cleanup

vmarta commented on packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts:
```diff
@@ -157,14 +149,14 @@ export class ApplicationExtension<
      * @throws Error if the deserialized component cannot be found in the component map.
      */
-    private deserialize(): DeserializedExtension | null {
+    private deserializeRoutes(): RouteProps[] | null {
```
There's a js error inside `deserializeRoutes`, on line 173: `const routes = serializedExtension.routes.map(...)`. Because the object shape for `window.__EXTENSIONS__` has changed and no longer has `routes` in it.

hajinsuha1 replied:
Ty! Updated the shape of the `window.__EXTENSIONS__` to include the `routes` field. https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2290/files#diff-152fe6a87add218e80ebf30f0f8ff4bd4242a7d406c90e1136fec332fffe6d66

vmarta commented on packages/extension-commerce-bm-seo/src/setup-app.ts:
```diff
@@ -0,0 +1,166 @@
+        const componentName =
+            config.resourceTypeToComponentMap[urlMapping.resourceType as keyof typeof config.resourceTypeToComponentMap]
+        
+        // Create a placeholder component so it can be deserialized later in beforeRouteMatch
```
Yes, but.. I think more importantly, we need a placeholder component because we don't have access to the component until later. This component is in a separate extension, that's why.

vmarta commented on packages/extension-commerce-bm-seo/src/setup-app.ts:
```diff
@@ -0,0 +1,166 @@
+            return Promise.resolve([])
+        }
+
+        // Transform URL mapping to serialized route
```
Maybe we don't need this comment anymore? We're not doing the serialization here in getRoutesAsync.

vmarta commented on packages/extension-commerce-bm-seo/src/setup-app.ts:
```diff
@@ -0,0 +1,166 @@
+        // and the values are placeholder components with the displayName
+        const {resourceTypeToComponentMap} = this.getConfig()
+        return Object.values(resourceTypeToComponentMap).reduce((acc: ComponentMap, name) => {
+            acc[`${this.getName()}.${name}`] = Object.assign(() => null, {displayName: `${this.getName()}.${name}`})
```
Yes, the component map in this case (SEO extension) returns _placeholder_ components. This may be obvious, but I thought it's quite eye opening that we're supposed to return placeholder components here :)

vmarta commented on packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts:
```diff
@@ -97,33 +96,28 @@ export class ApplicationExtension<
      * @throws Error if the routes have not been loaded.
      */
-    public serialize(): SerializedExtension | null {
+    public serializeRoutes(): SerializedRouteProps[] | null {
```
On naming this, how about `serializeAsyncRoutes`? To make it clearer that it may return null if there isn't any async route.

vmarta commented on packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts:
```diff
@@ -157,14 +149,14 @@ export class ApplicationExtension<
     protected getComponentMap?(): ComponentMap
```
Minor: how about making it slightly easier/more direct to get the component?

Previously, it's a 2-step operation:
1. `const componentMap = this.getComponentMap()`
2. `const component = componentMap[componentName]`

```suggestion
    protected getRouteComponent?(componentName: string): React.ComponentType<any>
```

hajinsuha1 replied:
Correct me if i'm wrong, but I believe this would make the implementation more complex for the extension developer. 

Wondering if it might be better to just have the extension developer implement returning the map while behind the scenes we handle the lookup?

vmarta reviewed (commented):
I think we're in the right direction with this PR 👍

vmarta commented on packages/pwa-kit-extension-sdk/src/types.ts:
```diff
@@ -46,13 +46,6 @@ export interface GetRoutesParams {
-/*
- * Same as react-router's RouteProps, but now supports `componentName` if the route is meant to be serializable
- */
-export type RouteProps = _RouteProps & {
```
FYI you'll need to replace all references to this RouteProps.. with the one from react-router instead.

hajinsuha1 committed:
address PR feedback

hajinsuha1 committed:
Merge branch 'feature/extensibility-v2' into implement-dynamic-update-routes

hajinsuha1 committed:
lint

hajinsuha1 committed:
support redirect url mappings

hajinsuha1 commented on packages/template-typescript-minimal/config/default.js:
```diff
@@ -0,0 +1,234 @@
+                    commerceAPIAuth: {
+                        propertyNameInLocals: 'commerceAPIAuth'
+                    },
+                    resourceTypeToComponentMap: {
+                        category: 'ProductList',
+                        product: 'ProductDetail',
+                        // TODO: what component should content be mapped to?
+                        content_asset: 'ProductList'
+                    }
```
@jeremy-jung1 here is where the config for the urlMapping resourceType is mapped to the component name. We could have this added to the commerce-bm-seo extension default.json

hajinsuha1 committed:
attempt passing props to ProductDetail/ProductList component

hajinsuha1 committed:
fix placeholder component

hajinsuha1 committed:
refactor to make setup-app more readable

hajinsuha1 committed:
implementation where PLP and PDP on client side doesn't get props

cc-prodsec commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/bd564fdb-a43e-4157-b784-5fa05d1d8237?refs=98a5011bfeafe4aa66df8c34983afa1745690cf3&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/bd564fdb-a43e-4157-b784-5fa05d1d8237/license?refs=98a5011bfeafe4aa66df8c34983afa1745690cf3&source=prChecksComment)

hajinsuha1 committed:
implement option 1 add componentProps to RouteProps and serialize it

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/setup-app.tsx:
```diff
@@ -0,0 +1,181 @@
+            {
+                path: requestURL.pathname,
+                component,
+                componentProps: props,
```
here we now pass the component's props as part of the route so it can be serialized

vmarta replied:
@hajinsuha1 just an aside: sometimes I feel you have important comments made in Github PR that I think it's best if they're written as code comments instead. Such comments in Github would be easily lost/forgotten.

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/setup-app.tsx:
```diff
@@ -0,0 +1,181 @@
+        }
+
+        // Create a map of the component names to a placeholder component for each unique name
+        return Array.from(new Set(Object.values(resourceTypeToComponentMap))).reduce(
+            (acc: ComponentMap, name) => {
+                acc[name] = createPlaceholderComponent(name)
+                return acc
+            },
+            {} as ComponentMap
+        )
```
This will create a map where the key's are the component names referenced in `resourceTypeToComponentMap` and the values are a placeholder component that has the displayName set to the component name

E.g., 
```
{
  ProductList: createPlaceholderComponent('ProductList'),
  ProductDetail: createPlaceholderComponent('ProductDetail'),
}
```

hajinsuha1 committed:
make deserialization work for both extension-seo and regular extensions

hajinsuha1 committed:
remove logging

hajinsuha1 committed:
allow extension-seo to work regardless of whether it is before/after extension-chakra-storefront

hajinsuha1 committed:
add AsyncRouteProps type

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/setup-app.tsx:
```diff
@@ -0,0 +1,174 @@
+     */
+    beforeRouteMatch({allRoutes}: BeforeRouteMatchParams): RouteProps[] {
+        const updatedRoutes = [...allRoutes]
+        const filteredRoutes = updatedRoutes.filter((route) => !route.component?.isPlaceholder)
```
we filter out any placeholder routes so that the SEO extension doesn't need to be placed before extension-chakra-storefront.

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/utils/utils.tsx:
```diff
@@ -0,0 +1,141 @@
+
+// getAppOrigin is going to be deprecated in PWA Kit v4. Currently we have a replacement (useOrigin) but it's a React hook. So we still need a non-hook version.
+// TODO: move to somewhere in SDK
+export const getAppOrigin = (
+    locals: Record<string, any> = {},
+    fromXForwardedHeader = false
+): string => {
+    if (typeof window !== 'undefined') {
+        return window.location.origin
+    }
+
+    const xForwardedOrigin = locals.xForwardedOrigin
+    if (fromXForwardedHeader && xForwardedOrigin) {
+        return xForwardedOrigin
+    }
+
+    const {APP_ORIGIN = ''} = process.env
+    return APP_ORIGIN
+}
```
@vmarta were you thinking this should be in `pwa-kit-react-sdk`?

hajinsuha1 committed:
remove getComponentName utility method

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/setup-app.tsx:
```diff
@@ -0,0 +1,172 @@
+
+            const componentProps = route.componentProps || {}
+
+            // We need to wrap the component with withPropsWrapper to ensure that the non-react
+            // specific statics are copied over such as `getComponent`
+            route.component = withPropsWrapper(actualComponent, componentProps)
```
we need to use `withPropsWrapper` HOC because we need to inject props directly into the component so it is rendered correctly

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/utils/utils.tsx:
```diff
@@ -0,0 +1,139 @@
+            ...commerceAPI.parameters,
+            proxy: `${appOrigin}${commerceAPI.proxyPath}`,
+            redirectURI: `${appOrigin}/callback`,
+            logger: console // TODO: proper logger
```
@vmarta any suggestions/example's currently in PWA kit for a proper logger here?

hajinsuha1 commented on packages/extension-starter/src/setup-app.ts:
```diff
@@ -79,14 +81,19 @@ class Sample extends ApplicationExtension<Config> {
-        return [
+    getRoutesAsync(params: GetRoutesParams): Promise<AsyncRouteProps[]> {
+        return Promise.resolve([
             {
                 exact: true,
                 path: this.getConfig().path,
-                component: SamplePage
+                component: SamplePage,
+                componentProps: {title: 'A custom title from the extension'}
             }
-        ]
+        ])
+    }
+
+    getComponentMap(): ComponentMap {
+        return {SamplePage}
```
here is an example of how regular extension's implementing `getRoutesAsync` would look like while passing props

hajinsuha1 commented on packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.tsx:
```diff
@@ -80,7 +83,7 @@ export class ApplicationExtension<
      * @returns a promise resolving to new routes to be added
      */
-    public getRoutesAsync?(params: GetRoutesParams): Promise<RouteProps[]>
+    public getRoutesAsync?(params: GetRoutesParams): Promise<AsyncRouteProps[]>
```
`getRoutesAsync` now returns a AsyncRouteProps type that contains `componentProps`

hajinsuha1 commented on packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.tsx:
```diff
@@ -178,14 +183,19 @@ export class ApplicationExtension<
                 )
             }
 
+            let component: React.ComponentType<any> = (props: Record<string, string>) => (
+                <Component {...props} {...route.componentProps} />
+            )
+            component = hoistNonReactStatics(component, Component)
+            component.displayName = componentName
```
when we deserialize we need to ensure
1. the serialized props are injected into the component
2. the non react statics are copied over to the wrapped component
3. the displayName is set correctly

hajinsuha1 committed:
cleanup

hajinsuha1 committed:
Merge branch 'feature/extensibility-v2' into implement-dynamic-update-routes

hajinsuha1 committed:
reorganize utils

hajinsuha1 committed:
support unit tests

hajinsuha1 committed:
add jest.config.js (copied from extension-chakra-store-locator)

hajinsuha1 committed:
add component-utils unit tests

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/utils/shopper-seo-utils.tsx:
```diff
@@ -0,0 +1,65 @@
+
+// getAppOrigin is going to be deprecated in PWA Kit v4. Currently we have a replacement (useOrigin) but it's a React hook. So we still need a non-hook version.
+// TODO: move to somewhere in SDK
+export const getAppOrigin = (
+    locals: Record<string, any> = {},
+    fromXForwardedHeader = false
+): string => {
+    if (typeof window !== 'undefined') {
+        return window.location.origin
+    }
+
+    const xForwardedOrigin = locals.xForwardedOrigin
+    if (fromXForwardedHeader && xForwardedOrigin) {
+        return xForwardedOrigin
+    }
+
+    const {APP_ORIGIN = ''} = process.env
+    return APP_ORIGIN
+}
```
@vmarta wanted to check if I still need to add `getAppOrigin` to `pwa-kit-react-sdk`?

vmarta replied:
@hajinsuha1 I think so. Currently getAppOrigin has this older implementation, and I'm not sure why it's being deprecated.
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/6cfecf0a0003e4834afab077661299c6fde8ed2a/packages/pwa-kit-react-sdk/src/utils/url.js#L23-L37

hajinsuha1 committed:
remove path from config and add shopper-seo-utils unit tests

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/types/config.ts:
```diff
@@ -0,0 +1,42 @@
+// Defines the map of resource types to component display name.
+// https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-seo?meta=getUrlMapping
+type ResourceTypeToComponentMap = {
+    category?: string
+    product?: string
+    content_asset?: string
```
i made these optional so that customers only need to define the resource types that they want to map in their app.
e.g., customer only uses `product` url mappings

hajinsuha1 commented on packages/extension-starter/src/pages/sample.tsx:
```diff
@@ -5,17 +5,23 @@
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 import React, {Fragment} from 'react'
+import PropTypes from 'prop-types'
```
these changes I made for testing purposes I will revert later.

hajinsuha1 commented on packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.tsx:
```diff
@@ -128,9 +131,11 @@ export class ApplicationExtension<
                 ...route,
-                componentName: route.component.displayName
+                componentName: route.component.displayName,
+                ...(route.componentProps && {componentProps: route.componentProps})
```
include `componentProps` in the serialized route if it exists in the route

vmarta replied:
@hajinsuha1 I'm seeing a typescript error on this line: 
```
Property 'componentProps' does not exist on type 'RouteProps<string, { [x: string]: string | undefined; }>'. (tsserver 2339)
```

Which makes the pwa-kit-extension-sdk package unable to build successfully.

hajinsuha1 replied:
that's interesting the return type of `serializeAsyncRoutes` should be `SerializedRouteProps` which contains `componentProps` maybe you don't have the latest code in your local? 
<img width="702" alt="Screenshot 2025-05-05 at 10 06 05 AM" src="https://github.com/user-attachments/assets/7efe4866-adf3-4026-8901-962ef8eb3edc" />

hajinsuha1 commented on packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.tsx:
```diff
@@ -178,14 +183,19 @@ export class ApplicationExtension<
                 )
             }
 
+            let component: React.ComponentType<any> = (props: Record<string, any>) => (
+                <Component {...props} {...route.componentProps} />
+            )
+            component = hoistNonReactStatics(component, Component)
+            component.displayName = componentName
```
1. render the component with it's props
2. include any non-react statics like `getComponent()`
3. set the display name to what it was before serialization

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/setup-app.tsx:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        // Make SEO GET Url Mapping API call
+        const urlSegment: string = locals.originalUrl.split('?')[0]
+        const config = this.getConfig()
+        const shopperSeo = await getShopperSeoClient(locals, config)
+        try {
+            urlMapping = await shopperSeo.getUrlMapping({
+                parameters: {urlSegment}
+            })
+        } catch (e) {
+            console.error(`Couldn't find mapping for given segement: ${urlSegment}`)
+        }
```
@yunakim714 here is where we make the SEO url mapping API call to handle server-side navigation

hajinsuha1 committed:
initial README.md

bendvc commented on packages/extension-commerce-bm-seo/.prettierignore:
```diff
@@ -0,0 +1,5 @@
+coverage
+dist
+generator-assets
+docs
+vendor
```
Are these all valid entries?

hajinsuha1 replied:
This was copied from `extension-starter` and I see that the other extensions have the same entries. However it does seem like these 3 entries don't exist

```
generator-assets
docs
vendor
```

bendvc commented on packages/extension-commerce-bm-seo/README.md:
````diff
@@ -0,0 +1,91 @@
+```sh
+npm install @salesforce/extension-commerce-bm-seo
+
+# Also: 
+# - install the peer dependencies listed in the package.json
+# - see the Peer Dependancies section below for any other steps (e.g. make sure your app uses CommerceApiProvider component)
+```
````
The peer dependencies automatically get installed. The fact that we have to do that when testing in our mono-repo is a side effect of lerna not being able to resolve peer dependencies correctly. So we don't really have to include this.

bendvc commented on packages/extension-commerce-bm-seo/README.md:
````diff
@@ -0,0 +1,91 @@
+# - see the Peer Dependancies section below for any other steps (e.g. make sure your app uses CommerceApiProvider component)
+```
+
+## Peer Dependencies
+
+PWA-Kit Application Extensions are NPM packages at their most simplest form, and as such you can define
+what peer dependencies are required when using it. Because this application extension provides
+Chakra UI via a page and components, it requires that the some peer dependencies are installed.
+
+Depending on what features your application extensions provides it's recommended you include any third-party
+packages as peer dependencies so that your base application doesn't end up having multiple versions of a 
+given package. See package.json for the full list of peer dependencies.
````
This this section required? I think this is still the boiler plate for the generated extension right?

hajinsuha1 replied:
removed!

bendvc commented on packages/extension-commerce-bm-seo/README.md:
```diff
@@ -0,0 +1,91 @@
+
+## Configuration
+
+The SEO extension is configured via the `mobify.app.extensions` property in your config files or `package.json`:
```
A lot of devs in our beta program where confused about where they can define their configuration. They assumed that it could only be defined in the package.json even tho we mention it being able to be defined in the config file. 

Maybe we can do something like:


```
The SEO extension is configured via the `app.extensions` property in your config (e.g. app/config/default.json) file or the key 'mobify.app/extensions' in your  `package.json`:
```

hajinsuha1 replied:
updated to 
```
The SEO extension is configured via your config file (e.g. `app/config/default.json`) or the key `mobify.app.extensions` in your `package.json`:
```

bendvc commented on packages/extension-commerce-bm-seo/README.md:
```diff
@@ -0,0 +1,91 @@
+          "@salesforce/extension-commerce-bm-seo",
+          {
+            "enabled": true,
+            "commerceAPI": {
```
We should explain that the commerceAPI is optional and the extension will use the provider used by other extensions if one exists. Like one from an extension configured before it.

bendvc commented on packages/extension-commerce-bm-seo/README.md:
```diff
@@ -0,0 +1,91 @@
+
+### Configuration Options
+
+- `commerceAPI`: Settings for connecting to the Commerce API
```
See above comment about this being optional

bendvc commented on packages/extension-commerce-bm-seo/README.md:
```diff
@@ -0,0 +1,91 @@
+
+The SEO extension works by:
+
+1. Intercepting incoming requests
```
I think we want to say that it block navigation instead of intercepting requests. We might also include that there will be a performance penalty using this extension so they know what they are getting into. It might be wise to put that somewhere closer to the top.

bendvc commented on packages/extension-commerce-bm-seo/config/default.json:
```diff
@@ -0,0 +1,19 @@
+        }
+    },
+    "commerceAPIAuth": {
+        "propertyNameInLocals": "commerceAPIAuth"
```
What is this?

hajinsuha1 replied:
this is the key used in locals for storing the `commerceAPIAuth`. @vmarta wondering if this needs to be in the config or we can just hardcode it in code?

vmarta replied:
@hajinsuha1 if you want to hardcode it, perhaps we should make the property name more unlikely to collide → something like `__commerceAPIAuth`.

bendvc commented on packages/extension-commerce-bm-seo/jest.config.js:
```diff
@@ -0,0 +1,27 @@
+    moduleNameMapper: {
+        ...base.moduleNameMapper,
+
+        // handle pwa-kit extensibility special import
+        '^overridable!(.*)': '$1'
```
No action required, but we might want to move this somewhere reusable so it doesn't show up in all the extensions. A simple solution would be to have a special config like '@salesforce/pwa-kit-dev/configs/jest/jest.config-extension.js')

bendvc commented on packages/extension-commerce-bm-seo/src/hooks/use-extension-store.ts:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {useApplicationExtensionsStore} from '@salesforce/pwa-kit-extension-sdk/react'
+import extensionMeta from '../../extension-meta.json'
+
+/**
+ * This hook returns the store for the current application extension.
+ */
+export const useExtensionStore = () => {
+    const state = useApplicationExtensionsStore(
+        (state: Record<string, any>) => state.state[extensionMeta.id]
+    )
+
+    if (state === undefined) {
+        throw new Error(
+            `'useExtensionStore' could not find your current application extension state! Make sure you have added the store to your extension.`
+        )
+    }
+
+    return state
+}
```
Are we using a store at all?

hajinsuha1 replied:
removed!

bendvc commented on packages/extension-commerce-bm-seo/src/setup-app.tsx:
```diff
@@ -0,0 +1,172 @@
+// Others
+import extensionMeta from '../extension-meta.json'
+
+interface StoreSlice {
+    count: number
+    increment: () => void
+    decrement: () => void
+}
+
+// This is safe to delete if your extension does not use state. If you aren't using this, ensure you remove the
+// `withApplicationExtensionStore` usage below as well.
+const sliceInitializer: SliceInitializer<StoreSlice> = (set) => ({
+    count: 0,
+    increment: () => set((state) => ({count: state.count + 1})),
+    decrement: () => set((state) => ({count: state.count - 1}))
+})
```
We can probably toss this is we aren't using it. I can't remember if Jer is using it or not tho.

bendvc commented on packages/extension-commerce-bm-seo/src/setup-app.tsx:
```diff
@@ -0,0 +1,172 @@
+    async getRoutesAsync({locals}: GetRoutesParams): Promise<AsyncRouteProps[]> {
+        let urlMapping
+
+        // TODO: do we need this check?
+        if (!locals.originalUrl) {
+            return Promise.resolve([])
+        }
```
That is most likely guaranteed to be there. So I'm not sure if it's required.But if we want a check, I'm wondering if we want to throw an error instead.

hajinsuha1 commented on packages/template-typescript-minimal/config/default.js:
```diff
@@ -0,0 +1,210 @@
+module.exports = {
+    app: {
+        extensions: [
+            ['@salesforce/extension-commerce-bm-seo', {}],
```
the order in which `extension-commerce-bm-seo` does not matter anymore for the server-side navigation to work

vmarta replied:
@hajinsuha1 But I think we'll want the seo extension to be first, though. So that other extensions after it can reuse the `Auth` object.

https://github.com/SalesforceCommerceCloud/pwa-kit/blob/6cfecf0a0003e4834afab077661299c6fde8ed2a/packages/extension-commerce-bm-seo/src/utils/shopper-seo-utils.tsx#L23-L31

hajinsuha1 commented on packages/extension-commerce-bm-seo/src/utils/shopper-seo-utils.tsx:
```diff
@@ -0,0 +1,65 @@
+// Local Imports
+import {Config} from '../types'
+
+export const getShopperSeoClient = async (locals: Record<string, any>, config: Config) => {
```
@bendvc @vmarta wanted to check in if we're okay going forward with this approach of pulling in `ShopperSeo` directly from `commerce-sdk-isomorphic`. This is because we need to make the API call in `getRoutesAsync` which is outside the HOC. On lines 24-31 it also looks like we handle saving the `Auth` in `locals` so we don't recreate it each time. 

I think this means the SEO extension *requires* the `commerceApi` to be defined in the config unlike `chakra-store-locator`

vmarta replied:
@hajinsuha1 I don't know if there's another way to do it. Our `commerce-sdk-react` does not provide us with a non-react way to get the api client. So we would have to rely on the underlying sdk-isomorphic lib.

To minimize the impact, we still wanted to use the Auth class from commerce-sdk-react, since that's one of the main benefits to using that library.

hajinsuha1 committed:
update README

vmarta reviewed (commented):
I'll continue to review this PR, in particular the files in the SEO extension.

vmarta commented on packages/extension-commerce-bm-seo/src/setup-app.tsx:
```diff
@@ -0,0 +1,172 @@
+     * is configured with, including those defined in the base application and those added by all the extensions. You can use this
+     * method to modify these routes in any way you want, but you must return an array of routes as a result.
+     */
+    beforeRouteMatch({allRoutes}: BeforeRouteMatchParams): RouteProps[] {
```
FYI there are several type errors in `beforeRouteMatch`:

![Ghostty 2025-05-02 at 16 33 52](https://github.com/user-attachments/assets/38fee81b-46f8-4e91-9f12-ed771e2f077d)

hajinsuha1 replied:
fixed!

vmarta commented on packages/extension-commerce-bm-seo/src/setup-app.tsx:
```diff
@@ -0,0 +1,172 @@
+            )?.component
+
+            if (!actualComponent) {
+                throw new Error(`Could not find component with displayName "${componentName}"`)
```
Minor: because `displayName` can be longer and does not match 1 to 1 with `componentName`:

```suggestion
                throw new Error(`Could not find component with displayName containing "${componentName}"`)
```

vmarta reviewed (commented):
Ok, done reviewing.

hajinsuha1 committed:
Merge branch 'feature/extensibility-v2' into implement-dynamic-update-routes

hajinsuha1 committed:
hardcode __commerceAPIAuth property in locals and use getAppOrigin from pwa-kit-react-sdk

hajinsuha1 committed:
cleanup

hajinsuha1 committed:
Merge branch 'implement-dynamic-update-routes' of https://github.com/SalesforceCommerceCloud/pwa-kit into implement-dynamic-update-routes

hajinsuha1 committed:
fix type errors

hajinsuha1 committed:
Revert to test changes

hajinsuha1 committed:
Revert "Revert to test changes"

This reverts commit ade541edd970fea80c373ba590890184baf8f8bf.

hajinsuha1 committed:
cleanup

hajinsuha1 committed:
REVERT TO TEST CHANGES

adamraya committed:
[App Extensibility ⚙️] Remove `extension-` prefix requirement (@W-18131976@) (#2390)

* Remove extension prefix

* Clean changes in template-typescript-minimal

* Optimize detecting extensions logic

* WIP Fix tests

* Fix tests

* Fix tests

* Move import constant

* Refactor packagePath ternary

* Condensed fullPackagePath into one line

* Update package namespace and name extract logic

* Use validate-npm-package-name in generator

* lint

* Adjust generator copy

* Update packages/pwa-kit-create-app/scripts/create-mobify-app.js

Co-authored-by: Kevin He <kevin.he@salesforce.com>
Signed-off-by: Adam Raya <adamraya@users.noreply.github.com>

* PR Feedback detectExtensions args

---------

Signed-off-by: Adam Raya <adamraya@users.noreply.github.com>
Co-authored-by: Kevin He <kevin.he@salesforce.com>

hajinsuha1 committed:
Merge branch 'feature/extensibility-v2' into implement-dynamic-update-routes

hajinsuha1 committed:
fix failing CI

hajinsuha1 committed:
fix failing CI

hajinsuha1 committed:
fix unit tests

hajinsuha1 committed:
fix unit tests

hajinsuha1 committed:
Merge branch 'feature/extensibility-v2' into implement-dynamic-update-routes

hajinsuha1 committed:
Merge branch 'implement-dynamic-update-routes' of https://github.com/SalesforceCommerceCloud/pwa-kit into implement-dynamic-update-routes

hajinsuha1 committed:
add comment for W-18474745 todo

hajinsuha1 committed:
Bump version to 0.1.0-extensibility-preview.5 and update dependencies in package.json

hajinsuha1 committed:
TEST CHANGES

hajinsuha1 committed:
Revert this commit to TEST CHANGES

This reverts commit 1d524d1be906f46ea53caf90a683919e89777729.
